dataRetriever/dataPool/headersCache/headersPool_test.go

```diff
@@ -0,0 +1,342 @@
+
+	hdrsCacher.Add(hdrHash1, testHdr1)
+	hdrsCacher.Add(hdrHash2, testHdr2)
+
+	hdr, err := hdrsCacher.GetHeaderByHash(hdrHash1)
+	assert.Nil(t, err)
+	assert.Equal(t, testHdr1, hdr)
+
+	hdr, err = hdrsCacher.GetHeaderByHash(hdrHash2)
+	assert.Nil(t, err)
+	assert.Equal(t, testHdr2, hdr)
+
+	expectedHeaders := []data.HeaderHandler{testHdr1, testHdr2}
+	hdrs, _, err := hdrsCacher.GetHeaderByNonceAndShardId(nonce, shardId)
+	assert.Nil(t, err)
+	assert.Equal(t, expectedHeaders, hdrs)
+}
+
+func TestHeadersCacher_AddHeadersInCacheAndRemoveByHash(t *testing.T) {
+	t.Parallel()
+
+	hdrsCacher, _ := headersCache.NewHeadersCacher(1000, 100)
+
+	nonce := uint64(1)
+	shardId := uint32(0)
+
+	hdrHash1 := []byte("hash1")
+	hdrHash2 := []byte("hash2")
+	testHdr1 := &block.Header{Nonce: nonce, ShardId: shardId}
+	testHdr2 := &block.Header{Nonce: nonce, ShardId: shardId, Round: 100}
+
+	hdrsCacher.Add(hdrHash1, testHdr1)
+	hdrsCacher.Add(hdrHash2, testHdr2)
+
+	hdrsCacher.RemoveHeaderByHash(hdrHash1)
+	hdr, err := hdrsCacher.GetHeaderByHash(hdrHash1)
+	assert.Nil(t, hdr)
+	assert.Equal(t, headersCache.ErrHeaderNotFound, err)
+
+	hdrsCacher.RemoveHeaderByHash(hdrHash2)
+	hdr, err = hdrsCacher.GetHeaderByHash(hdrHash2)
+	assert.Nil(t, hdr)
+	assert.Equal(t, headersCache.ErrHeaderNotFound, err)
+}
+
+func TestHeadersCacher_AddHeadersInCacheAndRemoveByNonceAndShadId(t *testing.T) {
+	t.Parallel()
+
+	hdrsCacher, _ := headersCache.NewHeadersCacher(1000, 100)
+
+	nonce := uint64(1)
+	shardId := uint32(0)
+
+	hdrHash1 := []byte("hash1")
+	hdrHash2 := []byte("hash2")
+	testHdr1 := &block.Header{Nonce: nonce, ShardId: shardId}
+	testHdr2 := &block.Header{Nonce: nonce, ShardId: shardId, Round: 100}
+
+	hdrsCacher.Add(hdrHash1, testHdr1)
+	hdrsCacher.Add(hdrHash2, testHdr2)
+
+	hdrsCacher.RemoveHeaderByNonceAndShardId(nonce, shardId)
+	hdr, err := hdrsCacher.GetHeaderByHash(hdrHash1)
+	assert.Nil(t, hdr)
+	assert.Equal(t, headersCache.ErrHeaderNotFound, err)
+
+	hdr, err = hdrsCacher.GetHeaderByHash(hdrHash2)
+	assert.Nil(t, hdr)
+	assert.Equal(t, headersCache.ErrHeaderNotFound, err)
+}
+
+func TestHeadersCacher_EvictionShouldWork(t *testing.T) {
+	t.Parallel()
+
+	hdrs, hdrsHashes := createASliceOfHeaders(1000, 0)
+	hdrsCacher, _ := headersCache.NewHeadersCacher(900, 100)
+
+	for i := 0; i < 1000; i++ {
+		hdrsCacher.Add(hdrsHashes[i], &hdrs[i])
+	}
+
+	// Cache will do eviction 2 times, in headers cache will be 800 headers
+	for i := 200; i < 1000; i++ {
+		hdr, err := hdrsCacher.GetHeaderByHash(hdrsHashes[i])
+		assert.Nil(t, err)
+		assert.Equal(t, &hdrs[i], hdr)
+	}
+}
+
+func TestHeadersCacher_ConcurrentRequestsShouldWorkNoEviction(t *testing.T) {
+	t.Parallel()
+
+	numHeadersToGenerate := 500
+
+	hdrs, hdrsHashes := createASliceOfHeaders(numHeadersToGenerate, 0)
+	hdrsCacher, _ := headersCache.NewHeadersCacher(numHeadersToGenerate+1, 10)
+
+	for i := 0; i < numHeadersToGenerate; i++ {
+		go func(index int) {
```

Comment:
Perhaps `Add` in a set of routines, `Get` in another set? More "real life" perhaps?

Reply:
done.